src/psbt.cpp

```diff
@@ -41,51 +47,227 @@ bool PartiallySignedTransaction::Merge(const PartiallySignedTransaction& psbt)
+    if (m_tx_modifiable == std::nullopt && psbt.m_tx_modifiable != std::nullopt) m_tx_modifiable = psbt.m_tx_modifiable;
     unknown.insert(psbt.unknown.begin(), psbt.unknown.end());
 
     return true;
 }
 
-bool PartiallySignedTransaction::AddInput(const CTxIn& txin, PSBTInput& psbtin)
+bool PartiallySignedTransaction::ComputeTimeLock(uint32_t& locktime) const
+{
+    std::optional<uint32_t> time_lock{0};
+    std::optional<uint32_t> height_lock{0};
+    for (const PSBTInput& input : inputs) {
+        if (input.time_locktime != std::nullopt && input.height_locktime == std::nullopt) {
+            height_lock.reset(); // Transaction can no longer have a height locktime
+            if (time_lock == std::nullopt) {
+                return false;
+            }
+        } else if (input.time_locktime == std::nullopt && input.height_locktime != std::nullopt) {
+            time_lock.reset(); // Transaction can no longer have a time locktime
+            if (height_lock == std::nullopt) {
+                return false;
+            }
+        }
+        if (input.time_locktime && time_lock != std::nullopt) {
+            time_lock = std::max(time_lock, input.time_locktime);
+        }
+        if (input.height_locktime && height_lock != std::nullopt) {
+            height_lock = std::max(height_lock, input.height_locktime);
+        }
+    }
+    if (height_lock != std::nullopt && *height_lock > 0) {
+        locktime = *height_lock;
+        return true;
+    }
+    if (time_lock != std::nullopt && *time_lock > 0) {
+        locktime = *time_lock;
+        return true;
+    }
+    locktime = fallback_locktime.value_or(0);
+    return true;
+}
+
+CMutableTransaction PartiallySignedTransaction::GetUnsignedTx() const
 {
-    if (std::find(tx->vin.begin(), tx->vin.end(), txin) != tx->vin.end()) {
+    if (tx != std::nullopt) {
+        return *tx;
+    }
+
+    CMutableTransaction mtx;
+    mtx.version = *tx_version;
+    bool locktime_success = ComputeTimeLock(mtx.nLockTime);
+    assert(locktime_success);
+    uint32_t max_sequence = CTxIn::SEQUENCE_FINAL;
+    for (const PSBTInput& input : inputs) {
+        CTxIn txin;
+        txin.prevout.hash = input.prev_txid;
+        txin.prevout.n = *input.prev_out;
+        txin.nSequence = input.sequence.value_or(max_sequence);
+        mtx.vin.push_back(txin);
+    }
+    for (const PSBTOutput& output : outputs) {
+        CTxOut txout;
+        txout.nValue = *output.amount;
+        txout.scriptPubKey = *output.script;
+        mtx.vout.push_back(txout);
+    }
+    return mtx;
+}
+
+uint256 PartiallySignedTransaction::GetUniqueID() const
+{
+    if (tx != std::nullopt) {
+        return tx->GetHash();
+    }
+
+    // Get the unsigned transaction
+    CMutableTransaction mtx = GetUnsignedTx();
+    // Compute the locktime
+    bool locktime_success = ComputeTimeLock(mtx.nLockTime);
+    assert(locktime_success);
```

Comment:
Is there a reason to compute the locktime again? It is done in the call to `GetUnsignedTx` just above.

(In 0cf1a9febb0205dadc03cd87a7af074e7b4cfd8c)

Reply:
Good point, removed.